### ISSUE TITLE
Wait for all promises to complete?

### ISSUE THREAD

**Issue by schneems** [ruby-concurrency/concurrent-ruby#575]
I'm chaining together a bunch of promises in a loop. After the loop is executed I need to wait for them all to finish. I would have thought that calling `wait!` on the parent promise would do this, but it doesn't, instead it only waits for the first promise in the chain to complete.

Code! I'm building up a bunch of promises in a loop:

```ruby
promise = nil
exporters.each do |exporter|
  if promise.nil?
    promise = Concurrent::Promise.new(executor: executor) { exporter.call }
  else
    promise.then { exporter.call }
  end
end
if promise
  promise.execute
  concurrent_exporters << promise
end
```

Then later I want to make sure they're all done:

```ruby
concurrent_exporters.each(&:wait!)
```

But that doesn't work. The top level promises all have `@state=:fulfilled` however the other promises are stuck at `@state=:pending`. What's the best way to wait for __all__ promises in a chain to be done and raise any errors?

**Comment by schneems** [comment]
Did figure out you can wait for the last promise, would still be nice to have a top level way of waiting for a promise chain to finish. I don't know when you would ever want to wait on only the top level promise and not the full chain.

Also seems silly to add all that extra code just to keep track of the last promise:

```ruby
exporters.each do |exporter|
  next if exporter.skip?(logger)

  if !environment.export_concurrent
    exporter.call
    next
  end

  if promise.nil?
    last_promise = promise = Concurrent::Promise.new(executor: executor) { exporter.call }
  else
    last_promise = promise.then { exporter.call }
  end
end

if promise
  promise.execute
  concurrent_exporters << last_promise
end
```

**Comment by mars** [comment]
This looks like the right approach. Promises don't know anything about their context, with exception of the standard function closure.

Are you catching errors somewhere, or do Ruby promises not have the error swallowing behavior like JavaScript?

**Comment by jdantonio** [comment]
@schneems The original intent was for these promises to be used similar to JavaScript, with the callbacks being the primary interaction. I added the state methods mainly to provide consistency with other abstractions in the library. Waiting an a set of promises wasn't a use I had ever considered. Normally a create an array of futures and iterate over the array, waiting on each individually. That approach isn't perfect, and what you suggest seems very reasonable.

I'm on my way to Brazil for RubyConf so I won't be able to do any work on this for several days, but it's something I'd like to discuss further. I'm also interested in input from @pitr-ch.

**Comment by pitr-ch** [comment]
This is tricky to support. How would you ever know that the tree of promises is complete? New children can be added at any moment. I believe that waiting on the last promise which captures the part the user is interested in is right approach (other children can be added freely without issues). For that reason I am closing the issue, please reopen if the solution is not satisfactory.

Regarding error handling, in current promises you can pass rescue block to `#then`.

**Comment by schneems** [comment]
Thanks for the rescue block advice, I'll look into propagating the errors so they could be raised by the last promise.

You could handle that problem of never knowing if the chain will be added to by not allowing new promises to be added. Basically calling freeze on the whole thing when you execute "wait_all!" (or whatever you would want to call it).

Is there a better data model or pattern for doing work concurrently where you need to chain together actions in a specific order and guarantee they're completed?

What's a good use case for promises if you're never waiting for them to end and not raising errors? i could see an in memory worker queue being a fit, anything else?

**Comment by ianks** [comment]
It seems like Promise.all could be a fitting use case here, no?

**Comment by pitr-ch** [comment]
> Thanks for the rescue block advice, I'll look into propagating the errors so they could be raised by the last promise.

In new promises `future { 'a' } .then{ |v| raise v }.then { never happens }.value!` will raise the error, I think the old promises are propagating in the same way, don't they?

> You could handle that problem of never knowing if the chain will be added to by not allowing new promises to be added. Basically calling freeze on the whole thing when you execute "wait_all!" (or whatever you would want to call it).

> Is there a better data model or pattern for doing work concurrently where you need to chain together actions in a specific order and guarantee they're completed?

It could work. The deep_freeze suffers the same problem though, it would not happen atomically. It's hard to document the behaviour then. 

On the other hand parents of a promise are always immutable so it fits naturally. Zipping all the promises you would like to wait for to a single promise is best option I believe. If it's a chain waiting on the last promise is enough, if it's a tree waiting on a promise zipping all leafs is enough.

> What's a good use case for promises if you're never waiting for them to end and not raising errors? i could see an in memory worker queue being a fit, anything else?

Yeah anything asynchronously executed is good use-case. The body of the promise can report the exception by other means (e.g. store it in a job's status field).

> It seems like Promise.all could be a fitting use case here, no?

Yeah. `zip` might be better since it'll give all the results as well. (This is a single thing in new API.)
